docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,354 @@
+
+A single replica is deployed, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
+
+A kube-apiserver's metrics response measures megabytes in size. To reduce the amount of network traffic, this proposal
+utilizes compressed HTTP responses when scraping metrics. Outside the scope of this proposal, a future enhancement is
+possible, in which a metrics filter component is added as a sidecar to each ShootKapi pod. Each such sidecar would
+mirror the metrics endpoint of its respective ShootKapi, with the only addition of support for filtering based on an
+HTTP request parameter (e.g. `GET /metrics?name=apiserver_request_total` would forward a `GET /metrics` request to the
+ShootKapi, and then would reduce the response to only `apiserver_request_total` counters, before passing it on to the
+caller). 
+
+### Element: New Custom Pod Metric for ShootKapis
+A new `shoot:apiserver_request_total:rate_by_pod` pod custom metric is made available for each ShootKapi pod on the
+seed. It is provided by the gardener-custom-metrics component. It is the rate of API requests per second, broken down
+by ShootKapi pod.
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
```

Comment:
For simplicity, I guess we can deploy the custom metrics component unconditionally and only control which scaling approach is used in the `kubeapiserver` component via the feature gate.

Reply:
The motivation is strictly economical, and only applies to HA seeds. Gardener-custom-metrics contributes to cross-AZ traffic (approx. 65KB/ShootKapi/minute), so you want if off if you're not using it.